lambdas/src/apis/collections/controllers/collections.ts

```diff
@@ -150,46 +204,117 @@ function buildUrn(protocol: string, contract: string, option: string): string {
-  return entities && entities.length > 0 && entities[0].metadata ? entities[0] : undefined
+  if (!(entities && entities.length > 0 && entities[0].metadata)) throw new ServiceError('Entity not found', 404)
+
+  return entities[0]
 }
 
-export function getBodyShapes(representations: WearableMetadataRepresentation[]) {
+function getBodyShapes(representations: WearableMetadataRepresentation[]): string[] {
   const bodyShapes = new Set<WearableBodyShape>()
   for (const representation of representations) {
     for (const bodyShape of representation.bodyShapes) {
+      // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
       bodyShapes.add(bodyShape.split(':').pop()!)
     }
   }
   return Array.from(bodyShapes)
 }
 
+function isValidSize(size: string): size is ValidSize {
+  return sizes[size] !== undefined
+}
+
+function getImagePath(root: string, imageRequest: ImageRequest): string {
+  return root + `/images/` + getImageRequestId(imageRequest) + '.png'
+}
+
+// Using this folder structure allow us to find and remove older versions of the same urn (entity)
+function getImageRequestId({ urn, hash, size, rarityBackground }: ImageRequest): string {
+  return `${urn}/${hash}/${rarityBackground ?? 'thumbnail'}-${size}`
+}
+
+// Delete all images that are not the latest version (same hash)
+async function pruneObsoleteImages(root: string, urn: string, hash: string) {
+  const existsFolder = fs.existsSync(root + `/images/` + urn + '/' + hash)
+  if (!existsFolder) await cleanFolder(root + `/images/` + urn)
+}
+
+async function cleanFolder(folderPath: string) {
+  await fs.promises.rm(folderPath, { recursive: true, force: true })
+}
+
+async function getImage(
+  client: SmartContentClient,
+  rootStorageLocation: string,
+  imageRequest: ImageRequest
+): Promise<Buffer> {
+  const path = getImagePath(rootStorageLocation, imageRequest)
+
+  // Check if the image is already in the cache, otherwise build and store it
+  fs.existsSync(path) || (await saveImage(client, rootStorageLocation, imageRequest))
+
+  return await sharp(path).toBuffer()
+}
+
+/**
+ * Fetch image from the content service and resized image is stored in the cache.
+ * When rarityBackground is present, it composes the image with the rarity background.
+ */
+async function saveImage(client: SmartContentClient, rootStorageLocation: string, imageRequest: ImageRequest) {
+  const image = await client.downloadContent(imageRequest.hash)
+  const shouldResize = sizes[imageRequest.size] !== DEFAULT_IMAGE_SIZE
+  let finalImage: sharp.Sharp
+  if (imageRequest.rarityBackground) {
+    const background = await getRarityBackground(imageRequest.rarityBackground, imageRequest.size)
+    const resizedImage = shouldResize
+      ? await sharp(image).resize(sizes[imageRequest.size], sizes[imageRequest.size]).toBuffer()
+      : image
+    finalImage = sharp(background).composite([{ input: resizedImage }])
```

Comment:
```suggestion
      ? await sharp(image).resize(sizes[imageRequest.size], sizes[imageRequest.size])
      : image
    finalImage = sharp(background).composite([{ input: resizedImage }])
```

Are those `.toBuffer()` avoidable? (here and inside getImage)

Asking because it may be faster to use stream based APIs if available.

Reply:
will check another options!

Reply:
Sharp API doesn't provide streams, I guess we can stream from the file once it's stored and avoid reading it again.

https://sharp.pixelplumbing.com/api-output

Reply:
in this case `input` receives a `string` (path) or `Buffer`, will check if we can avoid buffer conversions.